_posts/2020-03-15-method-s-arguments-pt-1.markdown

```diff
@@ -0,0 +1,372 @@
+
+```ruby
+def foo(arg1, arg2 = 'default value X', arg3 = 'default value Y', arg4)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+  puts "arg3 is: #{arg3.inspect}"
+  puts "arg4 is: #{arg4.inspect}"
+end
+
+foo(1, 2, 3, 4)
+arg1 is: 1
+arg2 is: 2
+arg3 is: 3
+arg4 is: 4
+
+foo(1, 2, 3)
+arg1 is: 1
+arg2 is: 2
+arg3 is: "default value Y"
+arg4 is: 3
+
+foo(1, 2)
+arg1 is: 1
+arg2 is: "default value X"
+arg3 is: "default value Y"
+arg4 is: 2
+```
+
+What Ruby seems to do is:
+- assign left most arguments to the first parameters
+- assign right most arguments to the last parameters
+- if not enough arguments, rise an error
+- if there are unused arguments, assign them to the optional arguments in order from left to right
+
+> I would recommend to avoid this combination unless it's something with a really specific requirement since it's not too intuitive and it's hard to follow.
+
+Having optional arguments not grouped will rise an error:
+
+```ruby
+def foo(arg1, arg2 = 'default value2', arg3, arg4 = 'default value4')
+end
+
+# SyntaxError ((irb):1: syntax error, unexpected '=', expecting ')')
+```
+
+# Optional arguments based on other arguments
+
+Default value for optional arguments can use the value of other arguments:
+
+```ruby
+def foo(arg1, arg2 = arg1 * 2)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo(1)
+# => arg1 is: 1
+#    arg2 is: 2
+
+foo(1,3)
+# => arg1 is: 1
+#    arg2 is: 3
+```
+
+This can even depend on the value of optional arguments:
+
+```ruby
+def foo(arg1 = 1, arg2 = arg1 * 2)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo
+# => arg1 is: 1
+#    arg2 is: 2
+
+foo(2)
+# => arg1 is: 2
+#    arg2 is: 4
+```
+
+The only requirement is that the arguments used in the default value was previously assigned, this won't work:
+
+```ruby
+def foo(arg1 = arg2 * 2, arg2 = 1)
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo
+# NameError (undefined local variable or method `arg2' for main:Object)
+```
+
+Also, we don't need to have simple operations like that for the default value, we can assign a function that will be called when assigning the values:
+
+```ruby
+def two_times(x)
+  x * 2
+end
+
+def foo(arg1 = 1, arg2 = two_times(arg1))
+  puts "arg1 is: #{arg1.inspect}"
+  puts "arg2 is: #{arg2.inspect}"
+end
+
+foo
+# => arg1 is: 1
+#    arg2 is: 2
+
+foo(5)
+# => arg1 is: 5
+#    arg2 is: 10 # it's multiplying 5, not 1, it uses our value, not te default one
+
+foo(5, 6)
+# => arg1 is: 5
+#    arg2 is: 6 # we are actually overriding that multiplication
+```
+
+# Variable arguments
+
+This type of optional positional arguments don't have a default value. They exist only if assigned and you can access them using a special array of arguments:
```

Comment:
```suggestion
This type of optional positional argument doesn't have a default value. It exists only if assigned and you can access it using a special array of arguments:
```

Reply:
How could I write is in plural? since it's always an array of arguments (it's a single parameter, but any number of arguments)

Reply:
"These types of optional positional arguments don't have default values." Then same as you had for the second sentence.